File: 33-search-in-rotated-sorted-array/33-search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& arr, int target) {
        
        int low = 0;
        int high  = arr.size()-1;
        int mid;
        
        while(low<=high){
            
            int mid = (low+ (high-low)/2);
            
            if(arr[mid]==target) return mid;
            
            
            // if left side is sorted
            if(arr[low]<=arr[mid]){
                
                if(target>=arr[low] and target<=arr[mid]){  // check if target is in left half
                    high = mid-1;
                }else{
                    low = mid + 1;
                }
                
            }else{
                // righht side sorted;
                
                if(target>=arr[mid] and target<=arr[high]){ //  check if it is in right half
                    low =mid+1;
                }else{
                    high = mid-1;
                }
                
            }
            
        }
        
        return -1;
        
    }
};

File: 33-search-in-rotated-sorted-array/33-search-in-rotated-sorted-array.py
class Solution:
    def search(self, arr: List[int], target: int) -> int:
        
        if len(arr)==1:
            if arr[0]==target:
                return 0
            else:
                return -1
        
        l = 0
        r = len(arr)-1
        
        
        while l<=r:
            
            mid = (l+r)//2
            # print(mid)
            
            if arr[mid]==target:
                return mid
            
            if arr[l]<=arr[mid]:
                # left part sorted
                if target > arr[mid] or target<arr[l]:
                    l = mid+1
                else:
                    r = mid-1
            
            else:
                # right sorted
                
                if target < arr[mid] or target > arr[r]:
                    r = mid-1
                else:
                    l = mid+1
            
            
        
        return -1
        